contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+API Server. In order to maintain a history of revisions to specification type 
+Objects, we only need to persist snapshots of the target Object states 
+contained in the specification type when they are revised.
+
+One approach would be to, for every specification type,  have a 
+corresponding History type. For example, we could introduce a StatefulSetHistory 
+object that aggregates a PodTemplateSpec and a slice of PersistentVolumeClaims.
+The StatefulSet controller could use this object to store point in time 
+snapshots of versions of StatefulSetSpecs. However, this requires that we 
+introduce a new History Kind for all current and future controllers. It has the 
+benefit of type safety, but, for this benefit, we trade generality.
+
+Another approach would be to use PodTemplate objects. This mechanisms provides 
+the desired generality, but it only provides for the recording of versions of
+PodTemplateSpecs (e.g. For StatefulSet, we can not use PodTemplates to 
+record revisions to PersistentVolumeClaims). Also, it introduces the potential 
+for overlapping histories for two Objects of different Kinds, with the same 
+`.Name` in the same Namespace. Lastly, it constrains the PodTemplate Kind from 
+evolving to fulfill its original intention.
+
+We propose an approach that has analogs with the approach taken by the 
+[Mesos](http://mesos.apache.org/) community. Mesos frameworks, which are in some 
+ways like Kubernetes controllers, are responsible for check pointing, 
+persisting, and recovering their own state. This problem is so common that 
+Mesos provides a ["State Abstraction"](https://github.com/apache/mesos/blob/master/include/mesos/state/state.hpp) 
+that allows frameworks to persist their state in either ZooKeeper or the 
+Mesos Replicate Log (A Multi-Paxos based state machine used by the Mesos 
+Masters). This State Abstraction is a mutable, durable dictionary where keys 
+and values are opaque strings. As controllers only need the capability to 
+persist an immutable point in time snapshot of target Object states to 
+implement a revision history, we propose to use an immutable durable dictionary, 
+and we introduce the AnyState object for this purpose.
+
+``` golang
+// AnyState implements an immutable dictionary of state data to allow for a 
+// point in time snapshot. Clients are responsible for serializing 
+// and deserializing the objects that contain their internal state and mapping 
+// them to a set of keys. 
+// Once an AnyState has been successfully created, it can not be updated. The 
+// API Server will fail validation of all update requests that attempt to mutate 
+// the Data field. AnyStates may, however, be deleted.
+type AnyState struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+    // Data is a map of keys to the serialized objects that comprise the state 
+    // snapshot.
+    // Each key must consist of alphanumeric characters, '-', '_' or '.'.
+    // +optional
+    Data map[string]string
```

Comment:
The intention is only to provide storage for history using a generic object that provides a JSON encoding allowing for re-application of the state data via a strategic merge patch as per the prior discussion with @erictune and @bgrant0607. **The intention is not to provide a generic key value store for controllers**.

Is your primary concern that it will be used for unintentional purposes?  If so, the construction above doesn't mitigate this. You could serialize arbitrary information into a RawExtension  as well.

If you feel the name "AnyState" is misleading with respect to the intention we can change it "ControllerHistory".